services/common/rs_server_common/stac_api_common.py

```diff
@@ -0,0 +1,619 @@
+# Copyright 2024 CS Group
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module to share common functionalities for validating / creating stac items"""
+import copy
+import json
+import traceback
+from abc import ABC, abstractmethod
+from contextlib import asynccontextmanager
+from dataclasses import dataclass
+from functools import wraps
+from pathlib import Path
+from typing import (
+    Any,
+    AsyncIterator,
+    Callable,
+    List,
+    Literal,
+    Optional,
+    Self,
+    Type,
+    Union,
+)
+
+import stac_pydantic
+import stac_pydantic.links
+import yaml
+from fastapi import HTTPException, Request, status
+from pydantic import BaseModel, Field, ValidationError
+from rs_server_common import settings
+from rs_server_common.utils.logging import Logging
+from rs_server_common.utils.utils import (
+    extract_eo_product,
+    odata_to_stac,
+    validate_inputs_format,
+)
+
+logger = Logging.default(__name__)
+
+
+class Queryables(BaseModel):
+    """
+    BaseModel used to describe queryable holder.
+    See: site-packages/pypgstac/migrations/pgstac.0.8.6.sql
+    """
+
+    id: str = Field("", alias="$id")
+    type: str = Field("object")
+    title: str = Field("STAC Queryables.")
+    schema: str = Field("http://json-schema.org/draft-07/schema#", alias="$schema")  # type: ignore
+    properties: dict[str, Any] = Field({})
+
+    class Config:  # pylint: disable=too-few-public-methods
+        """Used to overwrite BaseModel config and display aliases in model_dump."""
+
+        allow_population_by_field_name = True
+
+
+class QueryableField(BaseModel):
+    """BaseModel used to describe queryable item."""
+
+    type: str
+    title: str
+    format: Optional[str] = None
+    pattern: Optional[str] = None
+    description: Optional[str] = None
+    enum: Optional[List[str]] = None
+
+
+@dataclass
+class MockPgstac(ABC):
+    """
+    Mock a pgstac database for the services (adgs, cadip, ...) that use stac_fastapi but don't need a database.
+    """
+
+    service: Literal["adgs", "cadip"]
+
+    # Set by stac-fastapi
+    request: Request | None = None
+    readwrite: Literal["r", "w"] | None = None
+
+    # adgs or cadip function
+    all_collections: Callable = None
+    select_config: Callable = None
+    stac_to_odata: Callable = None
+    map_mission: Callable = None
+
+    # Is the service adgs or cadip ?
+    adgs: bool = False
+    cadip: bool = False
```

Comment:
Maybe we need to change this in future, since we have to add new services, and can't create a variable for each

Reply:
It's just an easy way to implement specific code like this: 
```python
        if self.adgs:
            stac_params["constellation"], stac_params["platform"] = mission
        if self.cadip:
            stac_params["platform"] = mission
```
of course we'll have to evolve this for future services.